### ISSUE TITLE
document design behind `/run/opengl-driver`

### ISSUE THREAD

**Issue by kirelagin** [NixOS/nixpkgs#6802]
There is an explanation of what is going on [here](https://github.com/NixOS/nixpkgs/blob/aa564bbac53fe8977f551b06c39458389747d3e7/pkgs/development/libraries/mesa/default.nix#L14) but I couldn’t find anything on _why_ is it this way.

Currently this adds extra burden on non-NixOS as the symlink has to be maintained. To make things even more complicated the actual destination of the symlink on NixOS is a package that merges multiple drivers and it is not in nixpkgs, it is generated inside a nixos module.

Why is it necessary to find drivers at runtime?

**Comment by vcunat** [comment]
Many users want to use other drivers than those from mesa (including myself). For example, nouveau is typically much worse than the binary nvidia driver (in most respects). I *think* that using a different driver also implies the need to use the libGL*, libvdpau, etc. versions distributed with it.

**Comment by vcunat** [comment]
I can't see anything actionable about this issue.

**Comment by stapelberg** [comment]
> I can't see anything actionable about this issue.

Can you supply a pointer to the documentation which the issue submitter asked for? If not, then why do you not consider writing the documentation an action item?

**Comment by vcunat** [comment]
But https://github.com/NixOS/nixpkgs/issues/6802#issuecomment-80347109 covers the point, I believe.

**Comment by stapelberg** [comment]
Okay, so let me ask a couple of specific follow-up questions which are still unclear to me:

> Many users want to use other drivers than those from mesa (including myself). For example, nouveau is typically much worse than the binary nvidia driver (in most respects).

Why does driver choice necessitate the additional symlink? I.e., why can my choice not just be another NixOS config option?

> I _think_ that using a different driver also implies the need to use the libGL*, libvdpau, etc. versions distributed with it.

Who could we ask to clarify this assumption?

**Comment by vcunat** [comment]
It is another NixOS option. The symlink management is automatic.

**Comment by vcunat** [comment]
I'm relatively certain about the assumption. I don't know anyone around nixos.org who would know with more certainty; maybe there's even noone.

**Comment by stapelberg** [comment]
For more context, I managed to solve the issue I was having: an old version of chrome (51) was still installed in root’s profile. Removing that version and instead installing chrome system-wide (via `/etc/nixos/configuration.nix`) solved the issue.

However, this behavior violates one of the assumptions I had about NixOS: I assumed that software, once installed, will keep working. This is clearly not the case if chrome breaks when I install a newer version of the mesa drivers.

So, let’s try to approach this issue from another angle: could we change the behavior so that we don’t run the risk of breaking installed software?

**Comment by vcunat** [comment]
The drivers are inherently impure; that's our design choice. @stapelberg: was this another instance of #16779?

**Comment by stapelberg** [comment]
Yes, that is precisely the error message I was getting. Strangely enough, I didn’t find that issue when googling the error message yesterday.

I feel the fact that drivers are impure should be better pointed out, this is the first time I read about it.

I have to say, I still don’t understand why drivers _have to be_ impure (do they?). Is this to avoid combinatorial explosion (having to build software against all different drivers) or are there any other reasons?

**Comment by vcunat** [comment]
Yes, primarily. We have at least three OpenGL implementations.

**Comment by stapelberg** [comment]
Can we do a back-of-the-envelope calculation as to how much computational/storage resources we’re saving by doing this (provided resource savings are the main motivation)? It would make it easier to understand this trade-off for end users — the fact that a crucial part of my system is impure is pretty disappointing to me :|.

**Comment by vcunat** [comment]
You can point it by hand to whatever version you like, even a fixed one, without having to recompile anything.

**Comment by stapelberg** [comment]
I don’t see how your comment addresses my question. I’m interested in having a pure system, not in switching drivers.

**Comment by vcunat** [comment]
You could hard-code the driver used during build-time, via a wrapper to the executables that sets `$LD_LIBRARY_PATH`. Then it would be "pure", i.e. always using the same libGL regardless of the system it's run on. If that system happened to be using a different type of in-kernel driver, such pure approach would cause problems, *I think*.

**Comment by dezgeg** [comment]
> Can we do a back-of-the-envelope calculation as to how much computational/storage resources we’re saving by doing this (provided resource savings are the main motivation)? It would make it easier to understand this trade-off for end users.

The build products of the proprietary drivers are most likely non-redistributable due to their licenses, so the users of them would need to build most of the apps from source.

**Comment by stapelberg** [comment]
> The build products of the proprietary drivers are most likely non-redistributable due to their licenses, so the users of them would need to build most of the apps from source.

I might be understanding the term “non-redistributable” wrong, but you don’t need to distribute the driver package itself to use it when compiling the remaining packages. So, in practice, users would need to build that one package, but not all reverse-dependencies.

> If that system happened to be using a different type of in-kernel driver, such pure approach would cause problems, I think.

What problems?

**Comment by yegortimoshenko** [comment]
> What problems?

The problem is that the (choice of) hardware that you run the operating system on is inherently impure, as in, if you change your video card, we can't emulate the old one to support packages that have been built with `libGL.so` that ships with the driver, meaning that packages that were compiled before that would break.

The idea of an operating system is to abstract away hardware differences, and if NixOS were to use OpenGL purely, it wouldn't do a good job at being an operating system.

If there had been such problem with any other device, say, network card, then swapping card with the new one would break all packages that have been built with the old card in mind. Fortunately, we don't have to do that, Linux streamlines most of driver madness.

**Comment by FRidh** [comment]
New issue related to this:
https://github.com/NixOS/nixpkgs/pull/60985

**Comment by FRidh** [comment]
Related RFC https://github.com/NixOS/rfcs/pull/121
